base/array.jl

```diff
@@ -1805,15 +1805,19 @@ julia> filter(isodd, a)
 """
 filter(f, As::AbstractArray) = As[map(f, As)::AbstractArray{Bool}]
 
-function filter!(f, a::Vector)
-    insrt = 1
+function filter!(f, a::AbstractVector)
+    idx = eachindex(a)
+    state = start(idx)
+    (i, state) = next(idx, state)
```

Comment:
Since you just need to return immediately when the input is empty, I don't see why `deletat!` enters the picture at all.